#### NYC Property Price Model - Linear Regresssion and RandomTree Regression
Input dataset:

https://www.kaggle.com/new-york-city/nyc-property-sales

Using the Apache Spark ML pipeline (python), build a model to predict price of property on the available features for New York City.

#### Load Data

In [3]:
# File location and type
file_location = "/FileStore/tables/nyc_rolling_sales-3aec7.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df)

_c0 borough neighborhood blg_class tax_class block lot ease_ment bldg_class address apt zip res_unts com_unts tot_unts land_sqft gross_sqft yr_built tax_class_sale bldg_class_sale price sale_date 4 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2A 392 6 C2 153 AVENUE B 10009 5 0 5 1633 6440 1900 2 C2 6625000 2017-07-19T00:00:00.000+0000 5 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2 399 26 C7 234 EAST 4TH STREET 10009 28 3 31 4616 18690 1900 2 C7 - 2016-12-14T00:00:00.000+0000 6 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2 399 39 C7 197 EAST 3RD STREET 10009 16 1 17 2212 7803 1900 2 C7 - 2016-12-09T00:00:00.000+0000 7 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2B 402 21 C4 154 EAST 7TH STREET 10009 10 0 10 2272 6794 1913 2 C4 3936272 2016-09-23T00:00:00.000+0000 8 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2A 404 55 C2 301 EAST 10TH STREET 10009 6 0 6 2369 4615 1900 2 C2 8000000 2016-11-17T00:00:00.000+0000 9 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2 405 16 C4 516 EAST 12TH STREET 10009 20 0 20 2581 9730 1900 2 C4 - 2017-07-20T00:00:00.000+0000 10 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2B 406 32 C4 210 AVENUE B 10009 8 0 8 1750 4226 1920 2 C4 3192840 2016-09-23T00:00:00.000+0000 11 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2 407 18 C7 520 EAST 14TH STREET 10009 44 2 46 5163 21007 1900 2 C7 - 2017-07-20T00:00:00.000+0000 12 1 ALPHABET CITY 08 RENTALS - ELEVATOR APARTMENTS 2 379 34 D5 141 AVENUE D 10009 15 0 15 1534 9198 1920 2 D5 - 2017-06-20T00:00:00.000+0000 13 1 ALPHABET CITY 08 RENTALS - ELEVATOR APARTMENTS 2 387 153 D9 629 EAST 5TH STREET 10009 24 0 24 4489 18523 1920 2 D9 16232000 2016-11-07T00:00:00.000+0000 14 1 ALPHABET CITY 08 RENTALS - ELEVATOR APARTMENTS 2 394 44 D7 181 AVENUE C 10009 30 4 34 4295 21328 1910 2 D7 - 2017-07-20T00:00:00.000+0000 15 1 ALPHABET CITY 08 RENTALS - ELEVATOR APARTMENTS 2B 400 21 D1 530-532 EAST 5TH STREET 10009 10 0 10 3717 12350 2009 2 D1 10350000 2016-10-17T00:00:00.000+0000 16 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3 STREET, 1C 10009 0 0 0 - - 1920 2 C6 1 2016-09-06T00:00:00.000+0000 17 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3 STREET, 1C 10009 0 0 0 - - 1920 2 C6 499000 2017-03-10T00:00:00.000+0000 18 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3 STREET, 3A 10009 0 0 0 - - 1920 2 C6 10 2017-04-28T00:00:00.000+0000 19 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3RD STREET, 5A 10009 0 0 0 - - 1920 2 C6 529500 2017-06-09T00:00:00.000+0000 20 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3 STREET, 2E 10009 0 0 0 - - 1920 2 C6 423000 2017-07-14T00:00:00.000+0000 21 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 46 C6 317 EAST 3RD STREET, 12 10009 0 0 0 - - 1925 2 C6 501000 2017-03-16T00:00:00.000+0000 22 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 49 C6 311 EAST 3RD STREET, 17 10009 0 0 0 - - 1920 2 C6 450000 2016-09-01T00:00:00.000+0000 23 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 49 C6 311 EAST 3RD, 22 10009 0 0 0 - - 1920 2 C6 510000 2017-08-17T00:00:00.000+0000 24 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2C 374 8 C6 702 EAST 5TH STREET, 5E 10009 0 0 0 - - 1900 2 C6 - 2016-09-07T00:00:00.000+0000 25 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 375 30 C6 742 EAST 6TH STREET, 1 10009 0 0 0 - - 1902 2 C6 350000 2017-08-30T00:00:00.000+0000 26 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 376 14 C6 244 EAST 7TH STREET 10009 24 0 24 4131 16776 1928 2 C6 11900000 2017-06-21T00:00:00.000+0000 27 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 376 14 C6 244 EAST 7TH STREET 10009 0 0 0 - - 1928 2 C6 11900000 2017-06-21T00:00:00.000+0000 28 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 376 16 C6 248 EAST 7TH STREET, 24 10009 0 0 0 - - 1928 2 C6 350000 2017-01-30T00:00:00.000+0000 29 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 376 16 C6 248 EAST 7TH STREET, 2 10009 0 0 0 - - 1928 2 C6 404000 2017-07-07T00:00:00.000+0000

#### Data Description  

###### Context
This dataset is a record of every building or building unit (apartment, etc.) sold in the New York City property market over a 12-month period (2016 & 2017).

###### Content
This dataset contains the location, address, type, sale price, and sale date of building units sold. A reference on the trickier fields:

BOROUGH: A digit code for the borough the property is located in; in order these are Manhattan (1), Bronx (2), Brooklyn (3), Queens (4), and Staten Island (5).
BLOCK; LOT: The combination of borough, block, and lot forms a unique key for property in New York City. Commonly called a BBL.
BUILDING CLASS AT PRESENT and BUILDING CLASS AT TIME OF SALE: The type of building at various points in time. See the glossary linked to below.
For further reference on individual fields see the Glossary of Terms. For the building classification codes see the Building Classifications Glossary: https://www1.nyc.gov/assets/finance/downloads/pdf/07pdf/glossary_rsf071607.pdf

Note that because this is a financial transaction dataset, there are some points that need to be kept in mind:

- Many sales occur with a nonsensically small dollar amount: $0 most commonly. These sales are actually transfers of deeds between parties: for example, parents transferring ownership to their home to a child after moving out for retirement.
- This dataset uses the financial definition of a building/building unit, for tax purposes. In case a single entity owns the building in question, a sale covers the value of the entire building. In case a building is owned piecemeal by its residents (a condominium), a sale refers to a single apartment (or group of apartments) owned by some individual.

###### Acknowledgements
This dataset is a concatenated and slightly cleaned-up version of the New York City Department of Finance's Rolling Sales dataset.
###### Inspiration
What can you discover about New York City real estate by looking at a year's worth of raw transaction records? Can you spot trends in the market, or build a model that predicts sale value in the future?

In [5]:
df.printSchema()

root
-- _c0: integer (nullable = true)
-- borough: integer (nullable = true)
-- neighborhood: string (nullable = true)
-- blg_class: string (nullable = true)
-- tax_class: string (nullable = true)
-- block: integer (nullable = true)
-- lot: integer (nullable = true)
-- ease_ment: string (nullable = true)
-- bldg_class: string (nullable = true)
-- address: string (nullable = true)
-- apt: string (nullable = true)
-- zip: integer (nullable = true)
-- res_unts: integer (nullable = true)
-- com_unts: integer (nullable = true)
-- tot_unts: integer (nullable = true)
-- land_sqft: string (nullable = true)
-- gross_sqft: string (nullable = true)
-- yr_built: integer (nullable = true)
-- tax_class_sale: integer (nullable = true)
-- bldg_class_sale: string (nullable = true)
-- price: string (nullable = true)
-- sale_date: timestamp (nullable = true)

In [6]:
# descriptions of data
df.describe().show()

+-------+------------------+------------------+------------------+--------------------+------------------+-----------------+------------------+---------+----------+--------------------+-----------------+------------------+------------------+-------------------+------------------+------------------+-----------------+-----------------+------------------+---------------+--------------------+
summary| _c0| borough| neighborhood| blg_class| tax_class| block| lot|ease_ment|bldg_class| address| apt| zip| res_unts| com_unts| tot_unts| land_sqft| gross_sqft| yr_built| tax_class_sale|bldg_class_sale| price|
+-------+------------------+------------------+------------------+--------------------+------------------+-----------------+------------------+---------+----------+--------------------+-----------------+------------------+------------------+-------------------+------------------+------------------+-----------------+-----------------+------------------+---------------+--------------------+
 count| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548| 84548|
 mean|10344.359878412262|2.9987581019066094| null| null| 1.651910272669626|4237.218976202867|376.22401476084593| null| null| null|646.4486449389639| 10731.99161423097|2.0252637554998345|0.19355868855561337| 2.249183895538629| 3941.675775353369|4045.707478572432|1789.322976297488| 1.657484505842835| null| 1276456.4977638705|
 stddev| 7151.779436344402|1.2897900492296468| null| null|0.8448206674575593|3568.263406631049| 658.1368139996122| null| null| null|3901.433564499641|1290.8791471461102|16.721037014467925| 8.713183368246833|18.972584432026725|41983.972705061635|35032.49052739487|537.3449934336519|0.8193412115061982| null|1.1405255345987352E7|
 min| 4| 1|AIRPORT LA GUARDIA|01 ONE FAMILY DWE...| | 1| 1| | |****** 95TH STREET| | 0| 0| 0| 0| - | - | 0| 1| A0| - |
 max| 26739| 5| WYCKOFF HEIGHTS|49 CONDO WAREHOUS...| 4| 16322| 9106| | Z9| ZEREGA AVENUE| `| 11694| 1844| 2261| 2261| 9996| 9992| 2017| 4| Z9| 9999999|
+-------+------------------+------------------+------------------+--------------------+------------------+-----------------+------------------+---------+----------+--------------------+-----------------+------------------+------------------+-------------------+------------------+------------------+-----------------+-----------------+------------------+---------------+--------------------+

#### Preliminary Data Summary
- There are 84548 rows in the data set
- According to Kaggle data summary, ease_ment is empty and the first column is irrelevant
- Some of the features in the dataset are uploaded with incorrect datatype inferred by databricks and needs to be updated, example price appears as string but should be an integer. 
- If there are duplicates, the rows need to be deleted from dataset
- Based on the description provided, the price column needs to be reviewed and several rows with no price or $0 price for property will need to be deleted. 
- There are also several high value properties which is the sale of entire building and price for these sales is very high, in billions of dollars, this data will also need to be deleted to remove outliers.
- Other data in the dataset needs to be explored

#### Dataset Updates and Analysis

In [9]:
# remove columns ease_ment as it is empty and _c0 is not useful

df = df.drop("ease_ment", "_c0")
display(df)

borough neighborhood blg_class tax_class block lot bldg_class address apt zip res_unts com_unts tot_unts land_sqft gross_sqft yr_built tax_class_sale bldg_class_sale price sale_date 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2A 392 6 C2 153 AVENUE B 10009 5 0 5 1633 6440 1900 2 C2 6625000 2017-07-19T00:00:00.000+0000 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2 399 26 C7 234 EAST 4TH STREET 10009 28 3 31 4616 18690 1900 2 C7 - 2016-12-14T00:00:00.000+0000 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2 399 39 C7 197 EAST 3RD STREET 10009 16 1 17 2212 7803 1900 2 C7 - 2016-12-09T00:00:00.000+0000 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2B 402 21 C4 154 EAST 7TH STREET 10009 10 0 10 2272 6794 1913 2 C4 3936272 2016-09-23T00:00:00.000+0000 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2A 404 55 C2 301 EAST 10TH STREET 10009 6 0 6 2369 4615 1900 2 C2 8000000 2016-11-17T00:00:00.000+0000 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2 405 16 C4 516 EAST 12TH STREET 10009 20 0 20 2581 9730 1900 2 C4 - 2017-07-20T00:00:00.000+0000 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2B 406 32 C4 210 AVENUE B 10009 8 0 8 1750 4226 1920 2 C4 3192840 2016-09-23T00:00:00.000+0000 1 ALPHABET CITY 07 RENTALS - WALKUP APARTMENTS 2 407 18 C7 520 EAST 14TH STREET 10009 44 2 46 5163 21007 1900 2 C7 - 2017-07-20T00:00:00.000+0000 1 ALPHABET CITY 08 RENTALS - ELEVATOR APARTMENTS 2 379 34 D5 141 AVENUE D 10009 15 0 15 1534 9198 1920 2 D5 - 2017-06-20T00:00:00.000+0000 1 ALPHABET CITY 08 RENTALS - ELEVATOR APARTMENTS 2 387 153 D9 629 EAST 5TH STREET 10009 24 0 24 4489 18523 1920 2 D9 16232000 2016-11-07T00:00:00.000+0000 1 ALPHABET CITY 08 RENTALS - ELEVATOR APARTMENTS 2 394 44 D7 181 AVENUE C 10009 30 4 34 4295 21328 1910 2 D7 - 2017-07-20T00:00:00.000+0000 1 ALPHABET CITY 08 RENTALS - ELEVATOR APARTMENTS 2B 400 21 D1 530-532 EAST 5TH STREET 10009 10 0 10 3717 12350 2009 2 D1 10350000 2016-10-17T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3 STREET, 1C 10009 0 0 0 - - 1920 2 C6 1 2016-09-06T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3 STREET, 1C 10009 0 0 0 - - 1920 2 C6 499000 2017-03-10T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3 STREET, 3A 10009 0 0 0 - - 1920 2 C6 10 2017-04-28T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3RD STREET, 5A 10009 0 0 0 - - 1920 2 C6 529500 2017-06-09T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 40 C6 327 EAST 3 STREET, 2E 10009 0 0 0 - - 1920 2 C6 423000 2017-07-14T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 46 C6 317 EAST 3RD STREET, 12 10009 0 0 0 - - 1925 2 C6 501000 2017-03-16T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 49 C6 311 EAST 3RD STREET, 17 10009 0 0 0 - - 1920 2 C6 450000 2016-09-01T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 373 49 C6 311 EAST 3RD, 22 10009 0 0 0 - - 1920 2 C6 510000 2017-08-17T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2C 374 8 C6 702 EAST 5TH STREET, 5E 10009 0 0 0 - - 1900 2 C6 - 2016-09-07T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 375 30 C6 742 EAST 6TH STREET, 1 10009 0 0 0 - - 1902 2 C6 350000 2017-08-30T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 376 14 C6 244 EAST 7TH STREET 10009 24 0 24 4131 16776 1928 2 C6 11900000 2017-06-21T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 376 14 C6 244 EAST 7TH STREET 10009 0 0 0 - - 1928 2 C6 11900000 2017-06-21T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 376 16 C6 248 EAST 7TH STREET, 24 10009 0 0 0 - - 1928 2 C6 350000 2017-01-30T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 376 16 C6 248 EAST 7TH STREET, 2 10009 0 0 0 - - 1928 2 C6 404000 2017-07-07T00:00:00.000+0000 1 ALPHABET CITY 09 COOPS - WALKUP APARTMENTS 2 376 19 C6 254 EAST 7TH STREET, 4 10009

In [10]:
# remove duplicate rows if any from dataset

from pyspark.sql import Row
df0 = df.dropDuplicates()
df0.count()

Out[6]: 83783

- Removed 765 duplicate rows from the dataset and updated the dataframe

#### Next Data Updates
- Update the datatype of various columns that are incorrect
- Update borough columns so it is easy to identify them for analysis
- Review and conduct various data analysis to determine if there is any interesting information
- Review and update price column and prepare for linear model creation
- Review other columns and make adjustments as needed to dataset

In [13]:
# change datatype of some features

from pyspark.sql.types import StringType, IntegerType, DateType, FloatType
from pyspark.sql.functions import unix_timestamp
from pyspark.sql.functions import to_date, col

df0 = df0.withColumn("borough", df0["borough"].cast(StringType()))
df0 = df0.withColumn("price", df0["price"].cast(IntegerType()))
df0 = df0.withColumn("land_sqft", df0["land_sqft"].cast(FloatType()))
df0 = df0.withColumn("gross_sqft", df0["gross_sqft"].cast(FloatType()))
df0 = df0.withColumn("price", df0["price"].cast(IntegerType()))
df0 = df0.withColumn("sale_date", to_date(col("sale_date"),"yyyy-MM-DD").cast(DateType()))
df0.printSchema()

root
-- borough: string (nullable = true)
-- neighborhood: string (nullable = true)
-- blg_class: string (nullable = true)
-- tax_class: string (nullable = true)
-- block: integer (nullable = true)
-- lot: integer (nullable = true)
-- bldg_class: string (nullable = true)
-- address: string (nullable = true)
-- apt: string (nullable = true)
-- zip: integer (nullable = true)
-- res_unts: integer (nullable = true)
-- com_unts: integer (nullable = true)
-- tot_unts: integer (nullable = true)
-- land_sqft: float (nullable = true)
-- gross_sqft: float (nullable = true)
-- yr_built: integer (nullable = true)
-- tax_class_sale: integer (nullable = true)
-- bldg_class_sale: string (nullable = true)
-- price: integer (nullable = true)
-- sale_date: date (nullable = true)

In [14]:
# update the column value based on data description

from pyspark.sql.functions import when

df1 = df0.withColumn("borough", when(df0["borough"] == '1', 'Manhattan').otherwise(df0['borough']))
df2 = df1.withColumn("borough", when(df1["borough"] == '2', 'Bronx').otherwise(df1['borough']))
df3 = df2.withColumn("borough", when(df2["borough"] == '3', 'Brooklyn').otherwise(df2['borough']))
df4 = df3.withColumn("borough", when(df3["borough"] == '4', 'Queens').otherwise(df3['borough']))
df5 = df4.withColumn("borough", when(df4["borough"] == '5', 'Staten Island').otherwise(df4['borough']))

In [15]:
# register spark SQL tables for analysis purposes

df5.createOrReplaceTempView("df_nyc")

In [16]:
%sql
Select count(*) from df_nyc

count(1) 83783

In [17]:
%sql
Select borough, count(*) from df_nyc group by borough;

borough count(1) Queens 26548 Brooklyn 23843 Staten Island 8296 Manhattan 18102 Bronx 6994

- Queens seems to have majority of the sales followed by Brooklyn and Manhattan.

In [19]:
%sql
Select tax_class, count(*) from df_nyc group by tax_class;

tax_class count(1) 1A 1440 3 3 1C 186 2A 2492 2B 799 1 38329 1B 1208 4 5960 738 2C 1907 2 30721

- Most of the data consists of sales for tax_class 1 and 2
- tax_class 3 seems to be absent from the dataset
- Below is the description of the Tax Classes from: https://www1.nyc.gov/assets/finance/downloads/pdf/07pdf/glossary_rsf071607.pdf

• Class 1: Includes most residential property of up to three units (such as one-,
two-, and three-family homes and small stores or offices with one or two
attached apartments), vacant land that is zoned for residential use, and most
condominiums that are not more than three stories.

• Class 2: Includes all other property that is primarily residential, such as
cooperatives and condominiums.

• Class 3: Includes property with equipment owned by a gas, telephone or electric
company.

• Class 4: Includes all other properties not included in class 1, 2, and 3, such as
offices, factories, warehouses, garage buildings, etc.

In [21]:
%sql
Select sale_date, count(*) from df_nyc group by sale_date;

sale_date count(1) 2017-08-11 261 2017-01-06 288 2016-10-03 146 2017-01-27 359 2017-02-26 7 2017-01-24 345 2016-11-08 330 2016-12-19 383 2017-06-29 541 2017-02-16 383 2017-07-31 359 2017-08-14 242 2017-08-18 192 2017-04-09 7 2017-03-28 347 2016-09-23 339 2017-02-28 458 2016-09-30 471 2017-08-10 337 2017-06-30 489 2017-08-19 8 2017-07-06 341 2017-01-30 382 2017-05-11 322 2016-09-15 439 2016-12-12 351 2016-10-07 345 2017-02-10 354 2017-04-25 312 2016-12-13 321 2016-09-26 329 2017-03-19 9 2017-05-26 271 2016-09-09 332 2017-01-04 276 2016-09-27 361 2017-06-28 435 2016-10-23 9 2017-04-03 267 2017-06-14 361 2017-08-22 252 2017-08-08 307 2017-04-29 17 2017-02-07 302 2016-10-20 355 2017-01-10 313 2017-03-17 345 2016-10-27 423 2017-05-10 320 2016-10-24 238 2016-11-14 334 2017-05-09 279 2017-01-21 19 2017-07-03 64 2017-04-07 365 2016-09-05 6 2017-07-30 4 2016-12-15 442 2017-06-08 347 2017-08-04 274 2016-11-24 9 2017-06-01 335 2017-01-05 327 2016-11-30 443 2016-11-25 33 2017-08-20 4 2017-04-02 5 2017-06-09 296 2017-04-12 196 2016-12-20 419 2017-04-11 213 2016-12-21 411 2016-11-19 16 2017-01-13 325 2017-04-20 319 2017-04-05 358 2017-02-08 341 2017-06-15 472 2017-03-13 278 2017-02-25 19 2016-12-22 522 2017-08-21 239 2017-05-21 8 2017-03-16 392 2017-04-27 457 2016-09-22 434 2016-10-17 264 2017-05-27 16 2017-02-01 315 2016-09-14 359 2017-01-26 339 2017-08-16 271 2017-01-01 30 2017-06-19 347 2017-04-18 242 2017-07-10 279 2017-06-13 376 2017-01-20 343 2017-08-29 143 2017-06-25 7 2016-10-30 6 2016-09-12 301 2016-10-06 506 2017-03-05 14 2017-07-15 20 2017-08-07 415 2017-06-07 330 2017-03-21 321 2017-07-26 300 2017-06-02 302 2017-03-24 333 2017-04-30 16 2016-11-15 343 2016-09-18 13 2016-12-03 10 2017-01-11 325 2016-09-08 373 2016-10-21 296 2017-02-12 2 2017-02-24 329 2017-01-17 350 2017-06-18 7 2016-09-01 392 2017-05-28 5 2016-11-21 386 2016-12-31 49 2017-05-05 319 2017-05-19 332 2016-09-25 6 2017-03-20 285 2016-10-29 20 2017-06-06 320 2017-04-15 16 2016-11-06 4 2017-08-06 4 2017-01-25 296 2017-01-19 342 2016-10-26 388 2016-09-02 182 2017-04-26 333 2017-07-11 289 2017-02-23 383 2016-12-30 277 2017-03-30 492 2016-12-17 15 2017-07-12 322 2017-08-17 277 2016-12-02 342 2017-03-29 368 2017-02-06 259 2017-01-23 294 2016-10-09 14 2017-04-08 14 2017-01-08 2 2016-11-07 315 2017-06-17 16 2016-12-18 3 2016-09-16 321 2016-11-04 336 2017-03-11 16 2016-12-29 392 2017-03-06 270 2017-04-23 4 2016-12-27 270 2017-02-22 279 2016-12-05 367 2017-03-26 13 2017-05-07 7 2017-02-05 6 2017-01-07 18 2016-09-13 320 2016-10-14 415 2017-01-12 393 2017-03-18 21 2016-09-24 16 2017-07-27 384 2017-02-17 328 2017-08-01 255 2016-10-11 301 2016-11-23 290 2017-05-22 322 2017-02-03 319 2017-01-31 426 2017-06-22 410 2017-08-03 333 2016-12-11 9 2017-06-03 11 2017-07-18 325 2017-06-10 16 2017-01-16 51 2017-06-11 7 2017-04-22 12 2016-11-18 350 2016-11-01 337 2017-08-28 151 2016-10-12 162 2016-11-28 295 2017-05-30 341 2017-03-07 300 2016-11-17 393 2017-06-21 353 2017-05-14 3 2017-02-27 357 2016-10-16 14 2016-09-03 8 2017-03-27 344 2016-10-22 19 2017-02-18 13 2016-12-04 6 2017-07-22 9 2017-08-27 3 2016-09-17 16 2017-08-31 90 2016-11-27 2 2016-10-13 377 2017-05-08 304 2017-06-26 384 2017-05-06 22 2016-12-06 308 2016-11-26 18 2017-01-09 290 2016-12-14 389 2016-11-02 284 2016-09-19 332 2016-10-10 100 2017-06-16 337 2016-10-19 349 2017-05-23 341 2017-07-23 4 2017-04-06 383 2017-05-03 283 2017-03-10 301 2016-11-10 446 2017-05-04 358 2017-02-21 292 2017-03-23 354 2017-03-03 324 2017-07-17 308 2017-05-02 294 2017-03-22 322 2017-06-05 274 2017-08-30 125 2017-07-09 2 2016-11-05 14 2016-12-10 18 2017-05-17 254 2016-09-10 24 2016-11-29 408 2017-04-24 328 2017-07-25 346 2017-06-23 347 2017-08-25 137 2017-03-02 328 2017-02-02 330 2016-12-09 342 2017-07-21 310 2017-01-22 8 2017-07-04 4 2017-05-24 380 2017-03-12 7 2017-07-24 296 2017-05-25 511 2016-12-16 422 2017-01-15 8 2017-07-05 201 2017-07-07 313 2017-08-15 287 2017-07-19 340 2017-05-18 392 2017-05-12 322 2017-08-09 296 2017-07-02 3 2016-09-06 248 201

- Based on the above plots, there is no specific seasonal pattern in the sale_date other than a weekly pattern. 
- On Sunday there are no property sales in NYC as expected. As the week progresses, sales of property in NYC increase and peaks on Wednesday, Thursday, Friday based on visual review of plot above only.

##### Price

- Review and update dataset using price column
- This column is the label for regression model and so must be scrutinized and dataset adjusted

In [24]:
%sql
Select price, count(*) from df_nyc group by price;

price count(1) 10460000 1 499000 137 431379 1 446160 9 135000 56 488000 17 1116250 1 615950 1 330799 1 83250 1 113000 8 702000 2 458185 2 2562125 1 419000 23 1967442 1 150300 1 28088 1 111300 3 271000 10 1078000 3 11261845 1 217500000 1 745601 2 5142162 3 1888000 4 78064 1 1267238 1 2823302 1 275400 4 117500 3 2717000 1 4900 1 3643243 1 351750 1 182391612 1 308930 1 55675000 1 237276 1 95080 1 745500 1 688070 1 250741 1 2511406 1 5625000 2 739309 1 582400 1 1662500 1 24755490 1 455661 1 2230227 1 1499373 1 45680625 1 2532011 1 1884000 1 1906962 1 323250 1 1068857 1 377208 1 1484608 1 586500 1 82529 1 1353152 1 13470930 1 497874 1 994054 1 1733910 1 1451922 1 911460 1 870000 90 3550000 15 450000 427 1205000 15 809000 6 466900 1 291000 5 4950000 7 552000 6 697000 7 358250 1 85000 29 1112787 2 213000 4 2365000 5 259869 1 398361 1 2647450 7 599999 3 756255 1 441000 2 1169000 2 12383705 1 659900 1 167750 1 457314 4 223250 3 18522529 1 375926 1 841000 1 2774731 1 45630 2 289250 1 719900 1 7742700 2 241079 1 748414 1 899210 1 506797 1 1208662 1 533500 1 980270 1 345000 111 145000 74 1450000 123 3375000 6 89882 1 779000 8 352750 2 212000 13 275908 1 772000 14 6414359 1 686000 6 30547 1 596936 1 1503596 1 107820 1 191100 1 8200000 4 326434 1 9999999 1 4358110 1 31500000 3 7540000 1 2798507 1 320853 1 947000 1 476500 3 3000 9 866239 1 450400 1 3936432 1 883901 1 1282984 2 629900 1 1227500 1 4913056 1 458700 1 327522 1 108500 1 35044 2 841900 1 968500 1 7104000 1 770400 1 2595000 1 1572500 2 1588950 1 1922500 1 2679500 1 828753 1 78046 1 378250 1 2273697 1 2099004 1 514216 2 5480618 1 538078 1 1279940 1 393750 1 610000 138 950000 301 659000 22 1505000 5 822436 1 1018324 1 190500 3 183204 1 1102460 1 389500 2 78000 2 5850000 13 1181170 3 3475000 5 114000 3 458000 17 636304 1 952981 1 15782875 2 613000 8 218517 1 1437500 2 3460000 1 286200 2 4575000 3 301200 1 1417343 1 582475 1 191500 2 1690500 1 390180 1 3317458 1 2054523 1 413500 1 1823000 1 864800 1 574500 1 742304 1 263692 1 14996374 1 16975000 1 471450 1 1555125 1 1063053 1 645570 1 271727 1 675000 204 410000 203 350000 344 415000 129 76000 2 1780000 17 1890467 1 1728988 1 6438317 1 539672 8 1680112 8 1064250 1 4000 7 1295677 1 9862645 1 6975012 4 1094618 3 701192 1 19975437 1 2096 1 504237 1 16900552 1 5510000 1 2875327 1 1747500 1 6500 3 679965 1 28392 1 756500 1 263777 1 283500 2 1157483 1 158594 1 1144250 1 1755000 4 4025498 1 751400 1 650904 1 171250 1 4340000 1 351296 1 2474347 1 1125174 1 4170000 1 1588800 1 412000 19 4000000 41 199468 1 748413 8 4760318 1 417000 16 3955000 1 585392 2 2038500 1 1411765 1 82500 2 1975000 18 5051332 1 39000 3 945900 1 474800 1 284750 1 1200500 1 3780207 1 450500 1 1712000 1 1524084 1 11546512 2 13513488 1 25355 3 1162574 1 216600 1 723500 1 775118 1 645500 1 429300 1 3529671 1 3775116 1 998000 48 505000 90 270750 3 629000 19 1240000 17 2125000 19 23000 6 297000 9 45839605 1 3728000 1 1123000 2 130200 1 146000 6 4302106 1 8251359 1 370265 1 513644 1 15453750 1 2665000 3 441079 1 180285 1 644000 3 238888 1 13425648 1 992698 1 1201535 2 203650 1 2821500 1 200500 1 387900 1 992750 1 2334000 1 327356 1 695197 1 877938 2 3878160 1 3675000 2 321065 1 328600 1 242346 1 500000 350 169000 19 435000 152 965000 52 27000 4 747000 6 23738428 1 58070000 1 608648 1 499900 19 177786 1 987739 1 91000 3 1578378 1 265488 1 1749048 1 4625000 2 4375000 1 341718 1 304114 1 573756 1 761659 1 280878 2 271890 1 553800 1 3805492 1 402464 1 1211000 1 993960 1 3538888 1 1712500 1 518289 3 2815589 1 1083900 1 401591 1 6516800 1 459820 1 2333250 1 545900 1 3595164 1 401397 1 483333 1 45394 1 1130257 1 260000 152 600767 10 506000 8 506666 1 391723 1 24800000 1 137341 1 2153598 1 1329499 1 505986 8 218923 1 2359794 1 216200 2 9272250 1 2197310 1 438712 1 16905700 1 67888 1 2064000 1 1339325 2 3420000 1 8349650 2 1849000 2 1667800 1 275500 2 861875 1 6801337 1 78919 1 1081381 1 3516373 1 10797637 1 385121 2 2492000 1 1018250 1 214134 1 1820516 1 849750 1 1821470 1 121806 1 15957

In [25]:
%sql
Select price, count(*) from df_nyc group by price;

price count(1) 10460000 1 499000 137 431379 1 446160 9 135000 56 488000 17 1116250 1 615950 1 330799 1 83250 1 113000 8 702000 2 458185 2 2562125 1 419000 23 1967442 1 150300 1 28088 1 111300 3 271000 10 1078000 3 11261845 1 217500000 1 745601 2 5142162 3 1888000 4 78064 1 1267238 1 2823302 1 275400 4 117500 3 2717000 1 4900 1 3643243 1 351750 1 182391612 1 308930 1 55675000 1 237276 1 95080 1 745500 1 688070 1 250741 1 2511406 1 5625000 2 739309 1 582400 1 1662500 1 24755490 1 455661 1 2230227 1 1499373 1 45680625 1 2532011 1 1884000 1 1906962 1 323250 1 1068857 1 377208 1 1484608 1 586500 1 82529 1 1353152 1 13470930 1 497874 1 994054 1 1733910 1 1451922 1 911460 1 870000 90 3550000 15 450000 427 1205000 15 809000 6 466900 1 291000 5 4950000 7 552000 6 697000 7 358250 1 85000 29 1112787 2 213000 4 2365000 5 259869 1 398361 1 2647450 7 599999 3 756255 1 441000 2 1169000 2 12383705 1 659900 1 167750 1 457314 4 223250 3 18522529 1 375926 1 841000 1 2774731 1 45630 2 289250 1 719900 1 7742700 2 241079 1 748414 1 899210 1 506797 1 1208662 1 533500 1 980270 1 345000 111 145000 74 1450000 123 3375000 6 89882 1 779000 8 352750 2 212000 13 275908 1 772000 14 6414359 1 686000 6 30547 1 596936 1 1503596 1 107820 1 191100 1 8200000 4 326434 1 9999999 1 4358110 1 31500000 3 7540000 1 2798507 1 320853 1 947000 1 476500 3 3000 9 866239 1 450400 1 3936432 1 883901 1 1282984 2 629900 1 1227500 1 4913056 1 458700 1 327522 1 108500 1 35044 2 841900 1 968500 1 7104000 1 770400 1 2595000 1 1572500 2 1588950 1 1922500 1 2679500 1 828753 1 78046 1 378250 1 2273697 1 2099004 1 514216 2 5480618 1 538078 1 1279940 1 393750 1 610000 138 950000 301 659000 22 1505000 5 822436 1 1018324 1 190500 3 183204 1 1102460 1 389500 2 78000 2 5850000 13 1181170 3 3475000 5 114000 3 458000 17 636304 1 952981 1 15782875 2 613000 8 218517 1 1437500 2 3460000 1 286200 2 4575000 3 301200 1 1417343 1 582475 1 191500 2 1690500 1 390180 1 3317458 1 2054523 1 413500 1 1823000 1 864800 1 574500 1 742304 1 263692 1 14996374 1 16975000 1 471450 1 1555125 1 1063053 1 645570 1 271727 1 675000 204 410000 203 350000 344 415000 129 76000 2 1780000 17 1890467 1 1728988 1 6438317 1 539672 8 1680112 8 1064250 1 4000 7 1295677 1 9862645 1 6975012 4 1094618 3 701192 1 19975437 1 2096 1 504237 1 16900552 1 5510000 1 2875327 1 1747500 1 6500 3 679965 1 28392 1 756500 1 263777 1 283500 2 1157483 1 158594 1 1144250 1 1755000 4 4025498 1 751400 1 650904 1 171250 1 4340000 1 351296 1 2474347 1 1125174 1 4170000 1 1588800 1 412000 19 4000000 41 199468 1 748413 8 4760318 1 417000 16 3955000 1 585392 2 2038500 1 1411765 1 82500 2 1975000 18 5051332 1 39000 3 945900 1 474800 1 284750 1 1200500 1 3780207 1 450500 1 1712000 1 1524084 1 11546512 2 13513488 1 25355 3 1162574 1 216600 1 723500 1 775118 1 645500 1 429300 1 3529671 1 3775116 1 998000 48 505000 90 270750 3 629000 19 1240000 17 2125000 19 23000 6 297000 9 45839605 1 3728000 1 1123000 2 130200 1 146000 6 4302106 1 8251359 1 370265 1 513644 1 15453750 1 2665000 3 441079 1 180285 1 644000 3 238888 1 13425648 1 992698 1 1201535 2 203650 1 2821500 1 200500 1 387900 1 992750 1 2334000 1 327356 1 695197 1 877938 2 3878160 1 3675000 2 321065 1 328600 1 242346 1 500000 350 169000 19 435000 152 965000 52 27000 4 747000 6 23738428 1 58070000 1 608648 1 499900 19 177786 1 987739 1 91000 3 1578378 1 265488 1 1749048 1 4625000 2 4375000 1 341718 1 304114 1 573756 1 761659 1 280878 2 271890 1 553800 1 3805492 1 402464 1 1211000 1 993960 1 3538888 1 1712500 1 518289 3 2815589 1 1083900 1 401591 1 6516800 1 459820 1 2333250 1 545900 1 3595164 1 401397 1 483333 1 45394 1 1130257 1 260000 152 600767 10 506000 8 506666 1 391723 1 24800000 1 137341 1 2153598 1 1329499 1 505986 8 218923 1 2359794 1 216200 2 9272250 1 2197310 1 438712 1 16905700 1 67888 1 2064000 1 1339325 2 3420000 1 8349650 2 1849000 2 1667800 1 275500 2 861875 1 6801337 1 78919 1 1081381 1 3516373 1 10797637 1 385121 2 2492000 1 1018250 1 214134 1 1820516 1 849750 1 1821470 1 121806 1 15957

In [26]:
%sql
Select price, count(*) from df_nyc group by price;

price count(1) 10460000 1 499000 137 431379 1 446160 9 135000 56 488000 17 1116250 1 615950 1 330799 1 83250 1 113000 8 702000 2 458185 2 2562125 1 419000 23 1967442 1 150300 1 28088 1 111300 3 271000 10 1078000 3 11261845 1 217500000 1 745601 2 5142162 3 1888000 4 78064 1 1267238 1 2823302 1 275400 4 117500 3 2717000 1 4900 1 3643243 1 351750 1 182391612 1 308930 1 55675000 1 237276 1 95080 1 745500 1 688070 1 250741 1 2511406 1 5625000 2 739309 1 582400 1 1662500 1 24755490 1 455661 1 2230227 1 1499373 1 45680625 1 2532011 1 1884000 1 1906962 1 323250 1 1068857 1 377208 1 1484608 1 586500 1 82529 1 1353152 1 13470930 1 497874 1 994054 1 1733910 1 1451922 1 911460 1 870000 90 3550000 15 450000 427 1205000 15 809000 6 466900 1 291000 5 4950000 7 552000 6 697000 7 358250 1 85000 29 1112787 2 213000 4 2365000 5 259869 1 398361 1 2647450 7 599999 3 756255 1 441000 2 1169000 2 12383705 1 659900 1 167750 1 457314 4 223250 3 18522529 1 375926 1 841000 1 2774731 1 45630 2 289250 1 719900 1 7742700 2 241079 1 748414 1 899210 1 506797 1 1208662 1 533500 1 980270 1 345000 111 145000 74 1450000 123 3375000 6 89882 1 779000 8 352750 2 212000 13 275908 1 772000 14 6414359 1 686000 6 30547 1 596936 1 1503596 1 107820 1 191100 1 8200000 4 326434 1 9999999 1 4358110 1 31500000 3 7540000 1 2798507 1 320853 1 947000 1 476500 3 3000 9 866239 1 450400 1 3936432 1 883901 1 1282984 2 629900 1 1227500 1 4913056 1 458700 1 327522 1 108500 1 35044 2 841900 1 968500 1 7104000 1 770400 1 2595000 1 1572500 2 1588950 1 1922500 1 2679500 1 828753 1 78046 1 378250 1 2273697 1 2099004 1 514216 2 5480618 1 538078 1 1279940 1 393750 1 610000 138 950000 301 659000 22 1505000 5 822436 1 1018324 1 190500 3 183204 1 1102460 1 389500 2 78000 2 5850000 13 1181170 3 3475000 5 114000 3 458000 17 636304 1 952981 1 15782875 2 613000 8 218517 1 1437500 2 3460000 1 286200 2 4575000 3 301200 1 1417343 1 582475 1 191500 2 1690500 1 390180 1 3317458 1 2054523 1 413500 1 1823000 1 864800 1 574500 1 742304 1 263692 1 14996374 1 16975000 1 471450 1 1555125 1 1063053 1 645570 1 271727 1 675000 204 410000 203 350000 344 415000 129 76000 2 1780000 17 1890467 1 1728988 1 6438317 1 539672 8 1680112 8 1064250 1 4000 7 1295677 1 9862645 1 6975012 4 1094618 3 701192 1 19975437 1 2096 1 504237 1 16900552 1 5510000 1 2875327 1 1747500 1 6500 3 679965 1 28392 1 756500 1 263777 1 283500 2 1157483 1 158594 1 1144250 1 1755000 4 4025498 1 751400 1 650904 1 171250 1 4340000 1 351296 1 2474347 1 1125174 1 4170000 1 1588800 1 412000 19 4000000 41 199468 1 748413 8 4760318 1 417000 16 3955000 1 585392 2 2038500 1 1411765 1 82500 2 1975000 18 5051332 1 39000 3 945900 1 474800 1 284750 1 1200500 1 3780207 1 450500 1 1712000 1 1524084 1 11546512 2 13513488 1 25355 3 1162574 1 216600 1 723500 1 775118 1 645500 1 429300 1 3529671 1 3775116 1 998000 48 505000 90 270750 3 629000 19 1240000 17 2125000 19 23000 6 297000 9 45839605 1 3728000 1 1123000 2 130200 1 146000 6 4302106 1 8251359 1 370265 1 513644 1 15453750 1 2665000 3 441079 1 180285 1 644000 3 238888 1 13425648 1 992698 1 1201535 2 203650 1 2821500 1 200500 1 387900 1 992750 1 2334000 1 327356 1 695197 1 877938 2 3878160 1 3675000 2 321065 1 328600 1 242346 1 500000 350 169000 19 435000 152 965000 52 27000 4 747000 6 23738428 1 58070000 1 608648 1 499900 19 177786 1 987739 1 91000 3 1578378 1 265488 1 1749048 1 4625000 2 4375000 1 341718 1 304114 1 573756 1 761659 1 280878 2 271890 1 553800 1 3805492 1 402464 1 1211000 1 993960 1 3538888 1 1712500 1 518289 3 2815589 1 1083900 1 401591 1 6516800 1 459820 1 2333250 1 545900 1 3595164 1 401397 1 483333 1 45394 1 1130257 1 260000 152 600767 10 506000 8 506666 1 391723 1 24800000 1 137341 1 2153598 1 1329499 1 505986 8 218923 1 2359794 1 216200 2 9272250 1 2197310 1 438712 1 16905700 1 67888 1 2064000 1 1339325 2 3420000 1 8349650 2 1849000 2 1667800 1 275500 2 861875 1 6801337 1 78919 1 1081381 1 3516373 1 10797637 1 385121 2 2492000 1 1018250 1 214134 1 1820516 1 849750 1 1821470 1 121806 1 15957

Based on the above three plots, "Price" data column seems to contain a lot of low values. This was also mentioned in the data description as follows: "Many sales occur with a nonsensically small dollar amount: $0 most commonly. These sales are actually transfers of deeds between parties: for example, parents transferring ownership to their home to a child after moving out for retirement." 

Need to remove all null priced properties and then $0 values properties from the dataset.

Based on the BoxPlot and Quantile plots, on the higher side of the scale, there are some significant outlier prices as well that will need to be adjusted.

In [28]:
# identify all null values for price

from pyspark.sql.functions import isnan
df5.filter(df5["price"].isNull()).count()


Out[10]: 14177

In [29]:
# remove 14,177 null values from dataframe

df6 = df5.filter(df5.price.isNotNull())
df6.count()

Out[11]: 69606

In [30]:
# review the descriptive stats including min, max for price
# to determine what to additional values will require removal
df6.describe("price").show()
df6.select("price").summary("count", "min", "25%", "75%", "max").show()

+-------+------------------+
summary| price|
+-------+------------------+
 count| 69606|
 mean|1248971.2631382353|
 stddev| 7782453.983199421|
 min| 0|
 max| 1040000000|
+-------+------------------+

+-------+----------+
summary| price|
+-------+----------+
 count| 69606|
 min| 0|
 25%| 230000|
 75%| 950000|
 max|1040000000|
+-------+----------+

After removing the null values and based on the descriptive summary of price column, almost 50% of the prices fall between $230,000 and $950,000.

A judgetment call of minimum reasonable price in NYC area needs to be determined and it would be fair to assume that it is $100,000 USD.

On the higher side of the price, there are several outliers but note that in the description it does indicates that the higher price values the value can go as high as say $5,000,000 USD. This would remove any extreme outlier values and preserve majority of the dataset.

In [32]:
df7 = df6.filter(df6.price>100000)
df8 = df7.filter(df7.price<5000000)
df8.describe("price").show()
df8.select("price").summary("count", "min", "25%", "75%", "max").show()

+-------+-----------------+
summary| price|
+-------+-----------------+
 count| 54579|
 mean|867277.7485113322|
 stddev|775694.7101083843|
 min| 100335|
 max| 4996841|
+-------+-----------------+

+-------+-------+
summary| price|
+-------+-------+
 count| 54579|
 min| 100335|
 25%| 398000|
 75%| 995000|
 max|4996841|
+-------+-------+

In [33]:
# register spark SQL tables for analysis purposes

df8.createOrReplaceTempView("df_nyc2")

In [34]:
%sql
Select price, count(*) from df_nyc2 group by price;

price count(1) 499000 137 431379 1 446160 9 135000 56 488000 17 1116250 1 615950 1 330799 1 113000 8 702000 2 458185 2 2562125 1 419000 23 1967442 1 150300 1 111300 3 271000 10 1078000 3 745601 2 1888000 4 1267238 1 2823302 1 275400 4 117500 3 2717000 1 3643243 1 351750 1 308930 1 237276 1 745500 1 688070 1 250741 1 2511406 1 739309 1 582400 1 1662500 1 455661 1 2230227 1 1499373 1 2532011 1 1884000 1 1906962 1 323250 1 1068857 1 377208 1 1484608 1 586500 1 1353152 1 497874 1 994054 1 1733910 1 1451922 1 911460 1 870000 90 3550000 15 450000 427 1205000 15 809000 6 466900 1 291000 5 4950000 7 552000 6 697000 7 358250 1 1112787 2 213000 4 2365000 5 259869 1 398361 1 2647450 7 599999 3 756255 1 441000 2 1169000 2 659900 1 167750 1 457314 4 223250 3 375926 1 841000 1 2774731 1 289250 1 719900 1 241079 1 748414 1 899210 1 506797 1 1208662 1 533500 1 980270 1 345000 111 145000 74 1450000 123 3375000 6 779000 8 352750 2 212000 13 275908 1 772000 14 686000 6 596936 1 1503596 1 107820 1 191100 1 326434 1 4358110 1 2798507 1 320853 1 947000 1 476500 3 866239 1 450400 1 3936432 1 883901 1 1282984 2 629900 1 1227500 1 4913056 1 458700 1 327522 1 108500 1 841900 1 968500 1 770400 1 2595000 1 1572500 2 1588950 1 1922500 1 2679500 1 828753 1 378250 1 2273697 1 2099004 1 514216 2 538078 1 1279940 1 393750 1 610000 138 950000 301 659000 22 1505000 5 822436 1 1018324 1 190500 3 183204 1 1102460 1 389500 2 1181170 3 3475000 5 114000 3 458000 17 636304 1 952981 1 613000 8 218517 1 1437500 2 3460000 1 286200 2 4575000 3 301200 1 1417343 1 582475 1 191500 2 1690500 1 390180 1 3317458 1 2054523 1 413500 1 1823000 1 864800 1 574500 1 742304 1 263692 1 471450 1 1555125 1 1063053 1 645570 1 271727 1 675000 204 410000 203 350000 344 415000 129 1780000 17 1890467 1 1728988 1 539672 8 1680112 8 1064250 1 1295677 1 1094618 3 701192 1 504237 1 2875327 1 1747500 1 679965 1 756500 1 263777 1 283500 2 1157483 1 158594 1 1144250 1 1755000 4 4025498 1 751400 1 650904 1 171250 1 4340000 1 351296 1 2474347 1 1125174 1 4170000 1 1588800 1 412000 19 4000000 41 199468 1 748413 8 4760318 1 417000 16 3955000 1 585392 2 2038500 1 1411765 1 1975000 18 945900 1 474800 1 284750 1 1200500 1 3780207 1 450500 1 1712000 1 1524084 1 1162574 1 216600 1 723500 1 775118 1 645500 1 429300 1 3529671 1 3775116 1 998000 48 505000 90 270750 3 629000 19 1240000 17 2125000 19 297000 9 3728000 1 1123000 2 130200 1 146000 6 4302106 1 370265 1 513644 1 2665000 3 441079 1 180285 1 644000 3 238888 1 992698 1 1201535 2 203650 1 2821500 1 200500 1 387900 1 992750 1 2334000 1 327356 1 695197 1 877938 2 3878160 1 3675000 2 321065 1 328600 1 242346 1 500000 350 169000 19 435000 152 965000 52 747000 6 608648 1 499900 19 177786 1 987739 1 1578378 1 265488 1 1749048 1 4625000 2 4375000 1 341718 1 304114 1 573756 1 761659 1 280878 2 271890 1 553800 1 3805492 1 402464 1 1211000 1 993960 1 3538888 1 1712500 1 518289 3 2815589 1 1083900 1 401591 1 459820 1 2333250 1 545900 1 3595164 1 401397 1 483333 1 1130257 1 260000 152 600767 10 506000 8 506666 1 391723 1 137341 1 2153598 1 1329499 1 505986 8 218923 1 2359794 1 216200 2 2197310 1 438712 1 2064000 1 1339325 2 3420000 1 1849000 2 1667800 1 275500 2 861875 1 1081381 1 3516373 1 385121 2 2492000 1 1018250 1 214134 1 1820516 1 849750 1 1821470 1 121806 1 1595790 1 1266000 1 308359 1 3655000 1 1788047 1 936225 1 1697740 1 611000 4 2405000 1 210000 140 405000 126 1378000 2 850000 309 130000 75 481000 8 569213 1 3252000 1 2340000 4 682228 1 271959 1 533360 1 404000 6 868000 13 273500 3 162920 5 248329 1 1298268 2 1052450 1 223900 1 778500 2 3895000 1 927000 8 3999000 2 413700 1 414225 1 3781000 1 872180 1 2011043 2 918303 1 912000 2 512900 1 300011 1 171573 1 583193 1 3299990 1 1022076 1 2128000 1 1881899 1 289674 1 3030000 1 379000 24 1465000 16 357000 8 662500 6 1163555 2 380500 1 1083660 1 751500 3 397500 5 1471197 1 136000 8 1126000 1 3971175 3 978500 1 1644473 2 624925 1 618500 1 1308451 4 775551 1 474552 4 2482902 1 4223025 1 478500 4 2784765 1 792500 

- After cleaning up the price above, the distribution seems to be normal as per box plot above
- Total dataset is now 54579

##### Yr_Built

In [37]:
%sql
Select yr_built, count(*) from df_nyc2 group by yr_built;

yr_built count(1) 1959 337 1896 4 1990 214 1903 43 1975 410 1977 107 1924 217 2003 284 2007 624 1892 4 1974 99 2015 1130 1889 2 1871 1 1927 322 1875 3 1955 1345 1890 41 2006 583 1978 84 1925 2760 1908 57 1961 605 2013 421 1942 90 1939 299 1944 29 1899 1073 1902 22 1922 106 1952 594 1956 398 1934 18 1988 278 1997 100 1994 70 1968 62 2014 893 1917 51 1847 1 1951 513 1930 3176 1855 1 1856 1 1938 121 1973 164 1979 123 1846 1 1946 51 1904 16 1947 99 1894 4 1901 1075 1800 25 1971 114 1950 2072 1913 80 1949 321 1966 182 2004 476 1991 89 1967 147 1969 138 1982 98 1865 2 1850 5 1881 4 1932 181 1957 372 1945 820 1996 170 1989 250 1943 6 1849 1 1914 76 1998 110 1845 4 1963 730 1910 2131 1965 983 1905 401 1985 315 1960 1756 1898 3 1926 349 1970 590 1921 101 2012 148 1918 30 1864 2 1893 2 1916 81 1987 429 1923 185 1941 165 1912 71 1907 35 2009 381 1911 72 2016 625 1936 77 1832 1 1920 3802 1995 85 1931 1240 1937 204 1919 11 1906 51 1895 4 1915 693 1909 55 1980 266 1883 1 1854 1 1953 305 1880 14 2001 176 1972 191 1958 350 1992 98 1983 139 2005 738 1954 286 1935 1065 1984 131 1964 542 2000 132 1981 88 2010 198 1986 301 2011 66 1976 22 2008 564 1940 1607 1870 12 1851 2 2017 1 1948 184 1999 137 1891 1 1962 672 0 3692 2002 307 1900 679 1929 337 1928 452 1933 59 1993 84 1882 2

- Looking at the data for year built, the top years when the properties sold in NYC were constructed as follows:

Year - Bldg Constructed

1910 - 2,131

1920 - 3,802

1925 - 2,760

1930 - 3,176

1940 - 1,607

1950 - 2,072

1960 - 1,756

- Overall, it is evident that the properties in NYC are relatively old in origin but they have been upgraded several times and at the end of the day location, location, location still applies to purchase of property in any city
- Note that the dataset does contain almost 3,692 properties with no information on year building was constructed

In [39]:
# remove 6,885 value with year_built value of '0'

df9 = df8.filter(df8.yr_built>0)
df9.count()

Out[15]: 50887

In [40]:
# add new column age_build using the yr_build column to df

df10 = df9.withColumn("age_built", 2017-col("yr_built"))

- After removing the yr_built = 0, the dataset is at 50,887 rows
- Converted the year_built with age_built which will be more useful/meaningful in the model

In [42]:
# review gross_sqft & land_sqft for null values

from pyspark.sql.functions import isnan
df10.filter((df10["gross_sqft"] == "") | df10["gross_sqft"].isNull() | isnan(df10["gross_sqft"])).count()

Out[17]: 17790

In [43]:
df10.filter((df10["land_sqft"] == "") | df10["land_sqft"].isNull() | isnan(df10["land_sqft"])).count()

Out[18]: 17641

- As we want to preserve as much data as possible, both "gross_sqft" and "land_sqft" will need to be imputed to ensure null values are replaced for regression modelling
- Imputation estimator will be used for completing missing values, either using the mean or the median of the columns in which the missing values are located (see below - Regression Modelling, Data Preparation for execution)

In [45]:
%sql
Select res_unts, count(*) from df_nyc2 group by res_unts;

res_unts count(1) 148 1 31 2 85 1 78 2 193 1 28 3 26 5 27 8 12 20 22 3 47 1 152 2 291 1 1 24234 13 7 6 338 16 31 3 2336 20 17 120 1 96 2 48 2 5 119 19 5 64 1 15 14 462 1 9 48 17 3 324 1 286 1 72 1 4 644 8 118 100 1 39 1 23 3 7 55 446 1 84 1 136 1 51 1 63 1 10 44 102 1 38 1 25 4 24 6 121 1 143 1 29 1 21 2 1844 1 32 3 60 1 56 1 58 1 11 12 68 1 14 11 42 1 2 9690 30 4 184 1 0 16745 18 5 36 2

In [46]:
%sql
Select com_unts, count(*) from df_nyc2 group by com_unts;

com_unts count(1) 126 2 2261 1 12 2 22 1 1 1511 6 16 3 86 5 17 15 1 9 2 35 1 4 38 8 4 23 1 7 5 10 1 25 1 21 1 32 1 11 1 14 1 2 261 0 52624

In [47]:
%sql
Select tot_unts, count(*) from df_nyc2 group by tot_unts;

tot_unts count(1) 31 2 78 2 193 1 34 1 126 2 2261 1 28 2 26 7 27 10 192 1 1866 1 12 22 22 2 47 1 152 2 1 24898 13 6 6 366 16 36 86 1 3 2737 20 16 295 1 120 1 96 2 48 3 5 179 19 5 64 1 15 15 462 1 9 57 17 4 324 1 286 1 72 1 35 1 4 708 8 137 100 1 39 1 23 4 7 75 446 1 84 1 136 1 51 1 63 1 10 56 102 1 38 1 25 5 24 6 121 1 143 1 29 1 21 3 32 5 60 1 56 1 58 1 11 14 150 1 68 1 14 9 2 9583 30 3 0 15554 18 8 36 2

- There seem to be outliers in residential, commercial and total units data that need to be deleted to ensure a good distribution
- Also the dataset is skewed towards 0, so tot_unts that have 0 value and tot_unts = res_unts + com_unts is incorrect will be removed

In [49]:
df11 = df10.filter(df10.tot_unts>0)
df12 = df10.filter(df11.tot_unts<500)
df13 = df12.filter(df12.tot_unts == df12.res_unts + df11.com_unts)
df13.count()

Out[19]: 50575

In [50]:
df13.createOrReplaceTempView("df_nyc3")

In [51]:
%sql
Select tot_unts, count(*) from df_nyc3 group by tot_unts;

tot_unts count(1) 31 2 78 2 193 1 34 1 126 2 28 2 26 7 27 10 192 1 12 22 22 2 47 1 152 2 1 22159 13 6 6 366 16 36 86 1 3 2731 20 16 295 1 120 1 96 2 48 3 5 179 19 5 64 1 15 15 462 1 9 57 17 4 324 1 286 1 72 1 35 1 4 708 8 135 100 1 39 1 23 4 7 75 446 1 84 1 136 1 51 1 63 1 10 55 102 1 38 1 25 5 24 6 121 1 143 1 29 1 21 3 32 5 60 1 56 1 58 1 11 14 150 1 68 1 14 8 2 9573 30 3 0 14311 18 8 36 2

- The boxplot of tot_unts now looks more usable with outliers and additional incorrect rows are deleted
- Total rows in dataset is: 50575

In [53]:
%sql
Select blg_class, count(*) from df_nyc3 group by blg_class;

blg_class count(1) 28 COMMERCIAL CONDOS 6 30 WAREHOUSES 122 35 INDOOR PUBLIC AND CULTURAL FACILITIES 12 37 RELIGIOUS FACILITIES 45 04 TAX CLASS 1 CONDOS 912 12 CONDOS - WALKUP APARTMENTS 441 01 ONE FAMILY DWELLINGS 12416 08 RENTALS - ELEVATOR APARTMENTS 20 29 COMMERCIAL GARAGES 149 09 COOPS - WALKUP APARTMENTS 2392 39 TRANSPORTATION FACILITIES 1 11A CONDO-RENTALS 10 16 CONDOS - 2-10 UNIT WITH COMMERCIAL UNIT 22 31 COMMERCIAL VACANT LAND 10 23 LOFT BUILDINGS 6 17 CONDO COOPS 920 41 TAX CLASS 4 - OTHER 9 14 RENTALS - 4-10 UNIT 285 02 TWO FAMILY DWELLINGS 9567 11 SPECIAL CONDO BILLING LOTS 1 21 OFFICE BUILDINGS 125 15 CONDOS - 2-10 UNIT RESIDENTIAL 799 32 HOSPITAL AND HEALTH FACILITIES 13 06 TAX CLASS 1 - OTHER 29 05 TAX CLASS 1 VACANT LAND 38 07 RENTALS - WALKUP APARTMENTS 1350 22 STORE BUILDINGS 355 27 FACTORIES 69 10 COOPS - ELEVATOR APARTMENTS 10978 26 OTHER HOTELS 9 13 CONDOS - ELEVATOR APARTMENTS 7205 38 ASYLUMS AND HOMES 7 33 EDUCATIONAL FACILITIES 12 03 THREE FAMILY DWELLINGS 2240

- There doesn't seem to be a need to do anything for blg_class = Building Class Category
- All values are complete and useful

#### Correlation

In [56]:
display(df13)

borough neighborhood blg_class tax_class block lot bldg_class address apt zip res_unts com_unts tot_unts land_sqft gross_sqft yr_built tax_class_sale bldg_class_sale price sale_date age_built Manhattan ALPHABET CITY 13 CONDOS - ELEVATOR APARTMENTS 2 373 1003 R4 324 EAST 4TH STREET 3A 10009 1 0 1 null null 1920 2 R4 762669 2016-09-21 97 Manhattan ALPHABET CITY 13 CONDOS - ELEVATOR APARTMENTS 2 402 1327 R4 100 AVENUE A 7A 10009 1 0 1 null null 2014 2 R4 2611811 2016-12-29 3 Manhattan ALPHABET CITY 15 CONDOS - 2-10 UNIT RESIDENTIAL 2C 372 1003 R1 296 EAST 2ND STREET D 10009 1 0 1 null null 2003 2 R1 805000 2016-09-06 14 Manhattan CHELSEA 10 COOPS - ELEVATOR APARTMENTS 2 770 59 D4 234 WEST 21ST STREET, 42 10011 0 0 0 null null 1913 2 D4 267701 2016-09-27 104 Manhattan CHELSEA 13 CONDOS - ELEVATOR APARTMENTS 2 798 1034 R4 170 WEST 23RD STREET 2Y 10011 1 0 1 null null 1973 2 R4 750000 2016-11-10 44 Manhattan FINANCIAL 13 CONDOS - ELEVATOR APARTMENTS 2 18 1326 R4 88 GREENWICH STREET 1611 10006 1 0 1 null null 1956 2 R4 960000 2017-06-30 61 Manhattan GRAMERCY 10 COOPS - ELEVATOR APARTMENTS 2 877 9 D4 60 GRAMERCY PARK NORTH, 2C 10010 0 0 0 null null 1928 2 D4 992793 2017-04-20 89 Manhattan GRAMERCY 10 COOPS - ELEVATOR APARTMENTS 2 899 1 D4 205 THIRD AVENUE, 5D 10003 0 0 0 null null 1964 2 D4 835000 2017-04-12 53 Manhattan GRAMERCY 17 CONDO COOPS 2 903 1127 R9 235 EAST 22ND STREET, 9G 10010 0 0 0 null null 1931 2 R9 972500 2017-08-14 86 Manhattan GREENWICH VILLAGE-CENTRAL 10 COOPS - ELEVATOR APARTMENTS 2 569 1 D4 43 5 AVENUE, PH11W 10003 0 0 0 null null 1905 2 D4 3800000 2017-01-25 112 Manhattan GREENWICH VILLAGE-WEST 02 TWO FAMILY DWELLINGS 1 619 19 B9 232 WEST 10TH STREET 10014 2 0 2 943.0 2684.0 1899 1 B9 3950000 2017-08-25 118 Manhattan GREENWICH VILLAGE-WEST 10 COOPS - ELEVATOR APARTMENTS 2 608 39 D4 101 WEST 12TH STREET, 19J 10011 0 0 0 null null 1961 2 D4 1970000 2017-06-27 56 Manhattan GREENWICH VILLAGE-WEST 10 COOPS - ELEVATOR APARTMENTS 2 608 53 D4 125 WEST 12TH STREET, 2B 10011 0 0 0 null null 1918 2 D4 999999 2017-08-17 99 Manhattan HARLEM-CENTRAL 02 TWO FAMILY DWELLINGS 1 2050 103 B9 27 HAMILTON TERRACE 10031 2 0 2 1343.0 3400.0 1899 1 B9 2000000 2017-06-06 118 Manhattan HARLEM-CENTRAL 03 THREE FAMILY DWELLINGS 1 1911 19 C0 129 WEST 126TH STREET 10027 3 0 3 2495.0 3335.0 2000 1 C0 2200000 2016-10-13 17 Manhattan HARLEM-CENTRAL 15 CONDOS - 2-10 UNIT RESIDENTIAL 2C 2034 1107 R1 221 WEST 148 STREET 4A 10039 1 0 1 null null 2008 2 R1 585000 2017-05-16 9 Manhattan HARLEM-UPPER 13 CONDOS - ELEVATOR APARTMENTS 2 2096 1044 R4 725 RIVERSIDE DRIVE 7E 10031 1 0 1 null null 1912 2 R4 357225 2017-07-14 105 Manhattan KIPS BAY 13 CONDOS - ELEVATOR APARTMENTS 2 928 1305 R4 340 EAST 23RD STREET 10D 10010 1 0 1 null null 2007 2 R4 1650000 2017-06-02 10 Manhattan LOWER EAST SIDE 10 COOPS - ELEVATOR APARTMENTS 2 315 1 D4 266 EAST BROADWAY, B1707 10002 0 0 0 null null 1960 2 D4 808000 2017-03-02 57 Manhattan MIDTOWN EAST 10 COOPS - ELEVATOR APARTMENTS 2 1346 14 D4 333 EAST 53RD, 12A 10022 0 0 0 null null 1930 2 D4 555000 2016-12-13 87 Manhattan MIDTOWN EAST 10 COOPS - ELEVATOR APARTMENTS 2 1362 41 D4 420 EAST 51ST STREET, 10G 10022 0 0 0 null null 1962 2 D4 445000 2017-07-27 55 Manhattan MIDTOWN WEST 10 COOPS - ELEVATOR APARTMENTS 2 1030 48 D4 230 CENTRAL PARK SOUTH, 4G 10019 0 0 0 null null 1938 2 D4 1200000 2017-06-06 79 Manhattan MIDTOWN WEST 13 CONDOS - ELEVATOR APARTMENTS 2 1032 1388 R4 350 WEST 42ND STREET 41C 10036 1 0 1 null null 2004 2 R4 1750000 2017-05-04 13 Manhattan MIDTOWN WEST 13 CONDOS - ELEVATOR APARTMENTS 2 1269 1162 R4 15 WEST 53RD STREET 32E 10019 1 0 1 null null 1982 2 R4 1790000 2017-05-01 35 Manhattan MURRAY HILL 10 COOPS - ELEVATOR APARTMENTS 2 916 55 D4 210 EAST 36TH STREET, 6D 10016 0 0 0 null null 1960 2 D4 361000 2016-10-20 57 Manhattan MURRAY HILL 10 COOPS - ELEVATOR APARTMENTS 2 917 17 D4 225 EAST 36TH STREET, 15B 10016 0 0 0 null null 1963 2 D4 367349 2017-05-03 54 Manhattan MURRAY HILL 13 CONDOS - ELEVATOR APARTM

- Based on the correlation map above, price is positively correlated with gross square feet and there is some degree of correlation between price and age_built
- The remainder of the features do not have any specific pattern that can be identified visually based on correlation map above

#### Regression Models

#### Feature Selection

See https://www1.nyc.gov/assets/finance/downloads/pdf/07pdf/glossary_rsf071607.pdf for detailed overview of the features.

For regression modelling not all features provided are useful. So I will be including the following features only:

  - "borough" - location can influence the price of property in NYC
  
  - "blg_class" - building class represents the type and size of property (https://www1.nyc.gov/assets/finance/jump/hlpbldgcode.html)
  
  - "res_unts" - number of residential units in the building
  
  - "com_unts" - number of commercial units in the building
  
  - "tot_unts" - number of total units in the building
  
  - "land_sqft", - sqft of land
  
  - "gross_sqft" - gross sqft of property
  
  - "yr_built">"building age" - converted to building age and may influence price
  
  - "price" - label value
  
  - "tax_class" - taxes on the property may infulence decision of purchase
  
I will not be including the following:  

  - "neighborhood" - covered with "borough"
  
  - "block" - covered with "borough"
  
  - "lot" - covered with "borough"
  
  - "bldg_class" - covered with "borough"
  
  - "address" - not relevant to pricing
  
  - "apt" - not relevant to pricing
  
  - "zip" - not relevant to pricing
  
  - "tax_class_sale" - covered with tax_class
  
  - "bldg_class_sale" - covered with blg_class (Building Class Category)
  
  - "sale_date" - irrelevant based on plot above

##### Data Preparation

In [61]:
# impute gross_sqft and land_sqft to complete the null values
# drop old columns of gross_sqft, land_sqft and yr_built as it was update to age_built
# dropping all columns that won't be used, "gross_sqft" + "land_sqft" original columns and "yr_built"

from pyspark.ml.feature import Imputer

imputer = Imputer(inputCols=["gross_sqft", "land_sqft"], outputCols=["gross_sqft_impt", "land_sqft_impt"])
model = imputer.fit(df13)
df14 = model.transform(df13)

df15 = df14.drop("neighborhood", "block", "lot", "bldg_class", "address", "apt", "zip", "tax_class_sale", "bldg_class_sale", "sale_date", "land_sqft", "gross_sqft", "yr_built")
display(df15)

borough blg_class tax_class res_unts com_unts tot_unts price age_built gross_sqft_impt land_sqft_impt Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 762669 97 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 2611811 3 2063.6543 2802.6152 Manhattan 15 CONDOS - 2-10 UNIT RESIDENTIAL 2C 1 0 1 805000 14 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 267701 104 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 750000 44 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 960000 61 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 992793 89 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 835000 53 2063.6543 2802.6152 Manhattan 17 CONDO COOPS 2 0 0 0 972500 86 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 3800000 112 2063.6543 2802.6152 Manhattan 02 TWO FAMILY DWELLINGS 1 2 0 2 3950000 118 2684.0 943.0 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 1970000 56 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 999999 99 2063.6543 2802.6152 Manhattan 02 TWO FAMILY DWELLINGS 1 2 0 2 2000000 118 3400.0 1343.0 Manhattan 03 THREE FAMILY DWELLINGS 1 3 0 3 2200000 17 3335.0 2495.0 Manhattan 15 CONDOS - 2-10 UNIT RESIDENTIAL 2C 1 0 1 585000 9 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 357225 105 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 1650000 10 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 808000 57 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 555000 87 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 445000 55 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 1200000 79 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 1750000 13 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 1790000 35 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 361000 57 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 367349 54 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 665000 29 2063.6543 2802.6152 Manhattan 17 CONDO COOPS 2 0 0 0 575000 57 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 3300000 12 2063.6543 2802.6152 Manhattan 09 COOPS - WALKUP APARTMENTS 2 0 0 0 507500 107 2063.6543 2802.6152 Manhattan 09 COOPS - WALKUP APARTMENTS 2 0 0 0 460323 117 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 4300000 101 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 2400000 54 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 442500 58 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 675000 53 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 475000 55 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 1588000 30 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 1380000 26 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 1115000 60 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 1077003 15 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 715000 91 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 325000 117 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 1995000 32 2063.6543 2802.6152 Manhattan 17 CONDO COOPS 2 0 0 0 2080000 52 2063.6543 2802.6152 Manhattan 17 CONDO COOPS 2 0 0 0 1400000 55 2063.6543 2802.6152 Manhattan 09 COOPS - WALKUP APARTMENTS 2 0 0 0 420000 107 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 3000000 103 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 815000 95 2063.6543 2802.6152 Manhattan 10 COOPS - ELEVATOR APARTMENTS 2 0 0 0 600000 69 2063.6543 2802.6152 Manhattan 13 CONDOS - ELEVATOR APARTMENTS 2 1 0 1 3035000 12 2063.6543 2802.6152 Ma

In [62]:
# convert categorical features into vectors for the model

from pyspark.ml.feature import StringIndexer, OneHotEncoder

indexer1 = StringIndexer(inputCol="borough", outputCol="borough_index")
df16 = indexer1.fit(df15).transform(df15)

encoder1 = OneHotEncoder(dropLast=False, inputCol="borough_index", outputCol="borough_vec")
df17 = encoder1.transform(df16)

indexer2 = StringIndexer(inputCol="blg_class", outputCol="blg_class_index")
df18 = indexer2.fit(df17).transform(df17)

encoder2 = OneHotEncoder(dropLast=False, inputCol="blg_class_index", outputCol="blg_class_vec")
df19 = encoder2.transform(df18)

indexer3 = StringIndexer(inputCol="tax_class", outputCol="tax_class_index")
df20 = indexer3.fit(df19).transform(df19)

encoder3 = OneHotEncoder(dropLast=False, inputCol="tax_class_index", outputCol="tax_class_vec")
df21 = encoder3.transform(df20)

df22 = df21.drop("borough", "blg_class", "tax_class", "borough_index", "blg_class_index", "tax_class_index")

display(df22)


res_unts com_unts tot_unts price age_built gross_sqft_impt land_sqft_impt borough_vec blg_class_vec tax_class_vec 1 0 1 762669 97 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 1 0 1 2611811 3 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 1 0 1 805000 14 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(9), List(1.0)) List(0, 10, List(2), List(1.0)) 0 0 0 267701 104 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 1 0 1 750000 44 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 1 0 1 960000 61 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 992793 89 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 835000 53 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 972500 86 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(7), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 3800000 112 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 2 0 2 3950000 118 2684.0 943.0 List(0, 5, List(2), List(1.0)) List(0, 34, List(2), List(1.0)) List(0, 10, List(0), List(1.0)) 0 0 0 1970000 56 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 999999 99 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 2 0 2 2000000 118 3400.0 1343.0 List(0, 5, List(2), List(1.0)) List(0, 34, List(2), List(1.0)) List(0, 10, List(0), List(1.0)) 3 0 3 2200000 17 3335.0 2495.0 List(0, 5, List(2), List(1.0)) List(0, 34, List(5), List(1.0)) List(0, 10, List(0), List(1.0)) 1 0 1 585000 9 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(9), List(1.0)) List(0, 10, List(2), List(1.0)) 1 0 1 357225 105 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 1 0 1 1650000 10 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 808000 57 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 555000 87 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 445000 55 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 1200000 79 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 1 0 1 1750000 13 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 1 0 1 1790000 35 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 361000 57 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 367349 54 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(1), List(1.0)) List(0, 10, List(1), List(1.0)) 1 0 1 665000 29 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 575000 57 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(7), List(1.0)) List(0, 10, List(1), List(1.0)) 1 0 1 3300000 12 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(3), List(1.0)) List(0, 10, List(1), List(1.0)) 0 0 0 507500 107 2063.6543 2802.6152 List(0, 5, List(2), List(1.0)) List(0, 34, List(4), Lis

In [63]:
from pyspark.ml.feature import VectorAssembler

vectorAssembler = VectorAssembler(inputCols = ["res_unts", "com_unts", "tot_unts", "age_built", "gross_sqft_impt", "land_sqft_impt", "borough_vec","blg_class_vec", "tax_class_vec"], outputCol = 'features')
df23 = vectorAssembler.transform(df22)
df23 = df23.select(['features', 'price'])
display(df23)

features price List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 97.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 762669 List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 3.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 2611811 List(0, 55, List(0, 2, 3, 4, 5, 8, 20, 47), List(1.0, 1.0, 14.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 805000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(104.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 267701 List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 44.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 750000 List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 61.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 960000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(89.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 992793 List(0, 55, List(3, 4, 5, 8, 12, 46), List(53.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 835000 List(0, 55, List(3, 4, 5, 8, 18, 46), List(86.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 972500 List(0, 55, List(3, 4, 5, 8, 12, 46), List(112.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 3800000 List(0, 55, List(0, 2, 3, 4, 5, 8, 13, 45), List(2.0, 2.0, 118.0, 2684.0, 943.0, 1.0, 1.0, 1.0)) 3950000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(56.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 1970000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(99.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 999999 List(0, 55, List(0, 2, 3, 4, 5, 8, 13, 45), List(2.0, 2.0, 118.0, 3400.0, 1343.0, 1.0, 1.0, 1.0)) 2000000 List(0, 55, List(0, 2, 3, 4, 5, 8, 16, 45), List(3.0, 3.0, 17.0, 3335.0, 2495.0, 1.0, 1.0, 1.0)) 2200000 List(0, 55, List(0, 2, 3, 4, 5, 8, 20, 47), List(1.0, 1.0, 9.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 585000 List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 105.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 357225 List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 10.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 1650000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(57.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 808000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(87.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 555000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(55.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 445000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(79.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 1200000 List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 13.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 1750000 List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 35.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 1790000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(57.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 361000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(54.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 367349 List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 29.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 665000 List(0, 55, List(3, 4, 5, 8, 18, 46), List(57.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 575000 List(0, 55, List(0, 2, 3, 4, 5, 8, 14, 46), List(1.0, 1.0, 12.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 3300000 List(0, 55, List(3, 4, 5, 8, 15, 46), List(107.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 507500 List(0, 55, List(3, 4, 5, 8, 15, 46), List(117.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 460323 List(0, 55, List(3, 4, 5, 8, 12, 46), List(101.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 4300000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(54.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 2400000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(58.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 442500 List(0, 55, List(3, 4, 5, 8, 12, 46), List(53.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 675000 List(0, 55, List(3, 4, 5, 8, 12, 46), List(55.0, 2063.654296875, 2802.615234375, 1.0, 1.0, 1.0)) 475000 List(0, 55, List(0, 2, 3, 4, 5,

In [64]:
# split the dataframe into 70% training dataframe and 30% testing dataframe

splits = df23.randomSplit([0.7, 0.3])
train_df = splits[0]
test_df = splits[1]

#### Linear Regression

In [66]:
from pyspark.ml.regression import LinearRegression

lr = LinearRegression(featuresCol = "features", labelCol="price", maxIter=10, regParam=0.3, elasticNetParam=0.8)
lr_model = lr.fit(train_df)

print("Coefficients: " + str(lr_model.coefficients))
print("Intercept: " + str(lr_model.intercept))

Coefficients: [730.8107361632726,1343.826839100649,849.1320317840932,452.97948369339963,2.3564533248892636,-0.0,-215306.17731114384,57179.221014169576,607436.3656862725,-366372.8600148238,-388397.3246281538,-30400.592922768585,-327091.0351623494,55058.9784116748,304761.5236034356,-481877.76646213257,215675.79694115292,931748.441948293,-345491.2534871966,-46119.717868832566,-243953.50096955575,-140506.4075545672,386577.2838339325,1037729.0790531067,417568.64572963666,315406.345521908,676072.1222041914,816199.00385937,60350.50802727458,285612.5944413784,-89064.54822822865,251355.15054208858,2134591.194406737,1473657.3632764113,223115.1390928751,1019014.901746586,337187.3413804896,492795.3658459915,-44636.218405331034,-1291494.2069401497,1244809.8985394523,154671.56612558672,-544969.7241266173,0.0,0.0,11983.929169922036,-52876.9008739702,368658.5246022767,-357361.9631436891,452987.46240300644,-99407.95113626296,265674.3250610659,376068.0698541298,193111.25930712136,0.0]
Intercept: 828569.1261324125

In [67]:
trainingSummary = lr_model.summary

print("RMSE: %f" % trainingSummary.rootMeanSquaredError)
print("r2: %f" % trainingSummary.r2)

RMSE: 614934.835729
r2: 0.343199

In [68]:
lr_predictions = lr_model.transform(test_df)
lr_predictions.select("prediction", "price", "features").show(5)

from pyspark.ml.evaluation import RegressionEvaluator
lr_evaluator = RegressionEvaluator(predictionCol="prediction", labelCol="price",metricName="r2")

print("R Squared (R2) on test data = %g" % lr_evaluator.evaluate(lr_predictions))

+-----------------+-------+--------------------+
 prediction| price| features|
+-----------------+-------+--------------------+
630039.5410987989| 668000|(55,[0,2,3,4,5,6,...|
635669.7996384562|1700000|(55,[0,2,3,4,5,6,...|
636893.1446873221| 768000|(55,[0,2,3,4,5,6,...|
638003.3799765934| 860000|(55,[0,2,3,4,5,6,...|
 643541.73683658| 170000|(55,[0,2,3,4,5,6,...|
+-----------------+-------+--------------------+
only showing top 5 rows

R Squared (R2) on test data = 0.33868

In [69]:
test_result = lr_model.evaluate(test_df)
rmse_lr = test_result.rootMeanSquaredError
print("Root Mean Squared Error (RMSE) on test data = %g" % rmse_lr)

Root Mean Squared Error (RMSE) on test data = 608377

#### Random Tree Regression

In [71]:
from pyspark.ml.regression import RandomForestRegressor

rf = RandomForestRegressor(featuresCol = "features", labelCol = "price")
rf_model = rf.fit(train_df)

rf_predictions = rf_model.transform(test_df)
rf_predictions.select("prediction", "price", "features").show(5)


+-----------------+-------+--------------------+
 prediction| price| features|
+-----------------+-------+--------------------+
584757.0864421983| 668000|(55,[0,2,3,4,5,6,...|
584757.0864421983|1700000|(55,[0,2,3,4,5,6,...|
578938.6250687792| 768000|(55,[0,2,3,4,5,6,...|
584757.0864421983| 860000|(55,[0,2,3,4,5,6,...|
575409.2553486832| 170000|(55,[0,2,3,4,5,6,...|
+-----------------+-------+--------------------+
only showing top 5 rows

In [72]:
rf_evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="rmse")
rmse_rf = rf_evaluator.evaluate(rf_predictions)

print("Root Mean Squared Error (RMSE) on test data = %g" % rmse_rf)

Root Mean Squared Error (RMSE) on test data = 587059

#### Decision Tree Regression

In [74]:
from pyspark.ml.regression import DecisionTreeRegressor

dt = DecisionTreeRegressor(featuresCol ="features", labelCol = "price")
dt_model = dt.fit(train_df)

dt_predictions = dt_model.transform(test_df)
dt_evaluator = RegressionEvaluator(
    labelCol="price", predictionCol="prediction", metricName="rmse")

rmse_dt = dt_evaluator.evaluate(dt_predictions)

print("Root Mean Squared Error (RMSE) on test data = %g" % rmse_dt)

Root Mean Squared Error (RMSE) on test data = 585252

#### Gradient-Boosted Tree Regression

In [76]:
from pyspark.ml.regression import GBTRegressor

gbt = GBTRegressor(featuresCol = "features", labelCol = "price", maxIter=10)
gbt_model = gbt.fit(train_df)

gbt_predictions = gbt_model.transform(test_df)
gbt_predictions.select("prediction", "price", "features").show(5)

+-----------------+-------+--------------------+
 prediction| price| features|
+-----------------+-------+--------------------+
 626811.698415494| 668000|(55,[0,2,3,4,5,6,...|
788595.4710309657|1700000|(55,[0,2,3,4,5,6,...|
 569965.611973863| 768000|(55,[0,2,3,4,5,6,...|
635848.6344012825| 860000|(55,[0,2,3,4,5,6,...|
 569965.611973863| 170000|(55,[0,2,3,4,5,6,...|
+-----------------+-------+--------------------+
only showing top 5 rows

In [77]:
gbt_evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="rmse")
rmse_gbt = gbt_evaluator.evaluate(gbt_predictions)

print("Root Mean Squared Error (RMSE) on test data = %g" % rmse_gbt)

Root Mean Squared Error (RMSE) on test data = 568684

In [78]:
print("SUMMARY")
print()
print("Root Mean Squared Error (RMSE) on test data for Linear Regression = %g" % rmse_lr)
print("Root Mean Squared Error (RMSE) on test data for Random Tree Regression = %g" % rmse_rf)
print("Root Mean Squared Error (RMSE) on test data for Decision Tree Regression = %g" % rmse_dt)
print("Root Mean Squared Error (RMSE) on test data for Gradient Boosted Tree Regression = %g" % rmse_gbt)

SUMMARY

Root Mean Squared Error (RMSE) on test data for Linear Regression = 608377
Root Mean Squared Error (RMSE) on test data for Random Tree Regression = 587059
Root Mean Squared Error (RMSE) on test data for Decision Tree Regression = 585252
Root Mean Squared Error (RMSE) on test data for Gradient Boosted Tree Regression = 568684

#### Summary

Linear Regression -  $608K

Random Tree Regression -  $587K

Decision Tree Regression -  $585K

Gradient Boosted Tree Regression -  $568K

The regression model for NYC property data remains signaificantly error prone but there was improvement based on subsequent usage of different statistical models: from Linear Regression $608K to Gradient Boosted Tree Regression - $568K. 

If there is to be improvement in the model, further work will be required in the following areas:

- Improvement in the data fields being collect, perhaps number of rooms in the property, etc.
- A lot of data cleansing was required, so perhaps the data can be more cautiously collected or even automated. It might also make sense to separate the data for residential properties from the commercial but this may be difficult as the two types of properties may exist in the same building
- Further data wrangling may definitely be reuquired of existing data
- Other types statistical models can also be tried to improve the modelling results, like classification modelling